hack/test-integration.sh

```diff
@@ -96,16 +101,37 @@ loop="${TIMES:-1}"
 # hack/test-integration.sh Template*
 # hack/test-integration.sh "(WatchBuilds|Template)"
 tests=( $(go run "${OS_ROOT}/hack/listtests.go" -prefix="${OS_GO_PACKAGE}/${package}.Test" "${testexec}" | grep -E "${1-Test}") )
+
+if [[ -n "${junit_report}" ]]; then
+	tmp_dir="/tmp/openshift/test-integration"
+	mkdir -p ${tmp_dir}
+	test_output_file="${tmp_dir}/intergration-test.log"
+	test_error_file="${tmp_dir}/test-go-err.log"
+	junit_report_file="${tmp_dir}/report.xml"
+	rm -rf ${test_output_file} ${test_error_file} ${junit_report_file}
+fi
 # run each test as its own process
 ret=0
+test_result="ok"
 pushd "${OS_ROOT}/${package}" &>/dev/null
+TESTSTARTTIME=$(date +%s%3N)
 for test in "${tests[@]}"; do
-	for((i=0;i<${loop};i+=1)); do
-		if ! (exectest "${test}" ${@:2}); then
-			ret=1
-		fi
-	done
+        for((i=0;i<${loop};i+=1)); do
+                if ! (exectest "${test}" ${@:2}); then
+                        ret=1
+                        test_result="FAIL"
+                fi
+        done
 done
+TESTENDTIME=$(date +%s%3N)
+TESTDURATION=$(($TESTENDTIME - $TESTSTARTTIME))
+test_result_status="${test_result}        github.com/openshift/origin/test/integration    $(($TESTDURATION / 1000)).$(($TESTDURATION % 1000))s"
+if [[ -n "${test_output_file}" ]]; then
+	echo "${test_result_status}" >> "${test_output_file}"
+	cat ${test_output_file} | junitreport --type gotest --suites nested --stream --roots "github.com/openshift/origin/test/integration" --output "${junit_report_file}"
```

Comment:
I would refactor `os::test::junit::generate_oscmd_report` to call something like `os::test::junit::generate_report` and write a new function `os::test::junit::generate_gotest_report` that you would trap just like we trap the `oscmd` stuff. Would be useful to share that with `hack/test-go.sh` as well.